doc/tutorial.md

```diff
@@ -0,0 +1,594 @@
+
+Note, `keepalived` is modified by `DPVS` project to support some specific parameters. The codes is resident in `tools/keepalived` and the executable is `bin/keepalived`. And `ospfd`/`sshd` is the standard version.
+
+Let's start from *Full-NAT* example first, it's not the easiest but really popular.
+
+<a id='fnat'/>
+
+# Full-NAT Mode
+
+<a id='simple-fnat'/>
+
+## Simple Full-NAT (two-arm)
+
+This is a simple example for FullNAT (`FNAT`), forwarding between two interfaces. Let's say one is WAN interface (`dpdk1`) and another is LAN interface (`dpdk0`).
+
+![fnat-two-arm](./pics/fnat-two-arm.png)
+
+The setting including:
+
+* *ip-addresses* and *routes* for DPDK LAN/WAN network.
+* *VIP* on WAN interface (`dpdk1`)
+* `FNAT` service (vip:vport) and related `RS`
+* `FNAT` mode need at least one *LIP* on LAN interface (`dpdk0`)
+
+```bash
+#!/bin/sh -
+
+# add VIP to WAN interface
+./dpip addr add 10.0.0.100/32 dev dpdk1
+
+# route for WAN/LAN access
+# add routes for other network or default route if needed.
+./dpip route add 10.0.0.0/16 dev dpdk1
+./dpip route add 192.168.100.0/24 dev dpdk0
+
+# add service <VIP:vport> to forwarding, scheduling mode is RR.
+# use ipvsadm --help for more info.
+./ipvsadm -A -t 10.0.0.100:80 -s rr
+
+# add two RS for service, forwarding mode is FNAT (-b)
+./ipvsadm -a -t 10.0.0.100:80 -r 192.168.100.2 -b
+./ipvsadm -a -t 10.0.0.100:80 -r 192.168.100.3 -b
+
+# add at least one Local-IP (LIP) for FNAT on LAN interface
+./ipvsadm --add-laddr -z 192.168.100.200 -t 10.0.0.100:80 -F dpdk0
+```
+
+And you can use the commands below to check what's just set:
+
+```bash
+$ ./dpip addr show
+inet 10.0.0.100/32 scope global dpdk1
+     valid_lft forever preferred_lft forever
+inet 192.168.100.200/32 scope global dpdk0
+     valid_lft forever preferred_lft forever sa_used 0 sa_free 1032176 sa_miss 0
+```
+
+```bash
+$ ./dpip route show
+inet 10.0.0.100/32 via 0.0.0.0 src 0.0.0.0 dev dpdk1 mtu 1500 tos 0 scope host metric 0 proto auto
+inet 192.168.100.200/32 via 0.0.0.0 src 0.0.0.0 dev dpdk0 mtu 1500 tos 0 scope host metric 0 proto auto
+inet 192.168.100.0/24 via 0.0.0.0 src 0.0.0.0 dev dpdk0 mtu 1500 tos 0 scope link metric 0 proto auto
+inet 10.0.0.0/16 via 0.0.0.0 src 0.0.0.0 dev dpdk1 mtu 1500 tos 0 scope link metric 0 proto auto
+```
+
+```bash
+$ ./ipvsadm  -ln
+IP Virtual Server version 0.0.0 (size=0)
+Prot LocalAddress:Port Scheduler Flags
+  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
+TCP  10.0.0.100:80 rr
+  -> 192.168.100.2:80             FullNat 1      0          0
+  -> 192.168.100.3:80             FullNat 1      0          0
+```
+
+```bash
+$ ./ipvsadm  -G
+VIP:VPORT            TOTAL    SNAT_IP              CONFLICTS  CONNS
+10.0.0.100:80        1
+                              192.168.100.200      0          0
+```
+
+And now have a try if FNAT (two-arm) works. I've set the Nginx server on RS (with TOA module) to response the real Client IP/Port directly as plain text (not html).
```

Comment:
And now verify if FNAT (two-arms) works. I've setup the Nginx server on RS (with TOA module) to response HTTP request of 10.0.0.100:80 as plain text (not html).